platform/core.startup/src/org/netbeans/core/startup/ModuleSystem.java

```diff
@@ -176,54 +176,69 @@ public void loadBootModules() {
-                    Exceptions.attachMessage(ioe, "URL: " + manifestUrl); // NOI18N
-                    throw ioe;
-                }
-                try {
-                    Manifest mani = new Manifest(is);
-                    Attributes attr = mani.getMainAttributes();
-                    if (attr.getValue("OpenIDE-Module") == null) { // NOI18N
-                        // Not a module.
+            
+            // process libs in 2 passes; first, process bootstrap libraries in platform/libs, creating
+            // FixedModules with a classloader that only loads from those libs.
+            // 2nd pass will process the remaining libraries, using this classloader.
+            do {
+                while (e.hasMoreElements()) {
+                    URL manifestUrl = e.nextElement();
+                    if (!checkedManifests.add(manifestUrl)) {
+                        // Already seen, ignore.
                         continue;
                     }
-                    bootModules.add(mgr.createFixed(mani, manifestUrl, loader));
-                } finally {
-                    is.close();
+                    URL jarURL = FileUtil.getArchiveFile(manifestUrl);
+                    if (jarURL != null && jarURL.getProtocol().equals("file") &&
+                            /* #121777 */ jarURL.getPath().startsWith("/")) {
+                        LOG.log(Level.FINE, "Considering JAR: {0}", jarURL);
+                    try {
+                            if (ignoredJars.contains(Utilities.toFile(jarURL.toURI()))) {
+                            LOG.log(Level.FINE, "ignoring JDK/JRE manifest: {0}", manifestUrl);
+                            continue;
+                        }
+                    } catch (URISyntaxException x) {
+                        Exceptions.printStackTrace(x);
+                    }
+                    }
+                    LOG.log(Level.FINE, "Checking boot manifest: {0}", manifestUrl);
+
+                    InputStream is;
+                    try {
+                        is = manifestUrl.openStream();
+                    } catch (IOException ioe) {
+                        // Debugging for e.g. #32493 - which JAR was guilty?
+                        Exceptions.attachMessage(ioe, "URL: " + manifestUrl); // NOI18N
+                        throw ioe;
+                    }
+                    try {
+                        Manifest mani = new Manifest(is);
+                        Attributes attr = mani.getMainAttributes();
+                        if (attr.getValue("OpenIDE-Module") == null) { // NOI18N
+                            // Not a module.
+                            continue;
+                        }
+                        bootModules.add(mgr.createFixed(mani, manifestUrl, loader));
+                    } finally {
+                        is.close();
+                    }
                 }
-            }
+                e = null;
+                if (upperE != null) {
```

Comment:
Wow. What a C-like coding trick! What about extracting the behavior into another method and calling it once for `loader` and once for `upperLoader`?

Only stylistic note...